### PR TITLE
feat(gitlab): add support for 'merge when pipeline succeeds'

### LINKED ISSUE #3265: Use GitLab's built-in automerge
**What would you like Renovate to be able to do?**

I want Renovate to set the `merge_when_pipeline_succeeds` flag in GitLab MRs it creates.

**Describe the solution you'd like**

GitLab supports [annotating a Merge Request to be merged automatically when the Pipeline succeeds](https://docs.gitlab.com/ce/user/project/merge_requests/merge_when_pipeline_succeeds.html) - this option is also made [available via the API](https://docs.gitlab.com/ce/api/merge_requests.html#accept-mr). I'd propose that Renovate sets this annotation when `automerge` is enabled in the config to avoid having to wait for the next Renovatebot run. This is helpful to reduce resource consumption in self-hosted situations where the bot is run on a schedule.

**Describe alternatives you've considered**

1. Run Renovate on a higher frequency - works, but increases load.

### PR DESCRIPTION
**Pull request by fgreinacher**
This commit adds (back) support for GitLab's 'merge when pipeline succeeds'
feature. This feature needs to be enabled by bot owners explicitely because
of an possible race condition in current GitLab versions.

Closes #3265

### CODE REVIEW

**Comment by fgreinacher** [comment]
This is basically a port of what I did in #3308 with some restructurings and a new global config option that makes the usage of this feature a bot-owner decision.

I'm not super happy with the naming but couldn't come up with something better. Maybe someone else has a good idea?

File: lib/config/definitions.ts [hunk 1]
```diff
@@ -1999,6 +1999,13 @@ const options: RenovateOptions[] = [
     type: 'boolean',
     default: false,
   },
+  {
+    name: 'allowMergeWhenPipelineSucceeds',
+    description: `Enable or disable usage of GitLab's "merge when pipeline succeeds" feature when automerging PRs`,
+    type: 'boolean',
+    default: false,
+    stage: 'global',
```
Comment (rarkins) [inline lib/config/definitions.ts]: Are you certain that we don't want this configurable per-repo?

File: lib/config/definitions.ts [hunk 2]
```diff
@@ -1999,6 +1999,13 @@ const options: RenovateOptions[] = [
     type: 'boolean',
     default: false,
   },
+  {
+    name: 'allowMergeWhenPipelineSucceeds',
```
Comment (rarkins) [inline lib/config/definitions.ts]: I'd prefer to have GitLab in the option name, so users don't have to read the description to know that. Maybe `useGitLabAutomerge` ? Cc @ViceIce @JamieMagee

File: lib/workers/pr/index.js [hunk 3]
```diff
@@ -253,6 +253,8 @@ async function ensurePr(prConfig) {
         const platformOptions = {
           azureAutoComplete: config.azureAutoComplete,
           statusCheckVerify: config.statusCheckVerify,
+          mergeWhenPipelineSucceeds:
+            config.automerge && config.automergeType === 'pr',
```
Comment (rarkins) [inline lib/workers/pr/index.js]: Shouldn't this come from the config?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/definitions.ts (hunk 1 above)
Comment (fgreinacher) [reply lib/config/definitions.ts]: I think yes, for the reason described in https://github.com/renovatebot/renovate/issues/3265#issuecomment-536226275. Would you as the owner of the Renovate app on GitLab.com want to get complaints when people activate this in their repos and it fails?

File: lib/workers/pr/index.js (hunk 3 above)
Comment (fgreinacher) [reply lib/workers/pr/index.js]: With the current approach setting the `allowMergeWhenPipelineSucceeds` option is independent of the existing `automerge` option. To get something automerged, enable `automerge` , to gain a bit speed, additionallyy enable MWPS.

File: lib/config/definitions.ts (hunk 2 above)
Comment (fgreinacher) [reply lib/config/definitions.ts]: That sounds better, will update!

File: lib/config/definitions.ts (hunk 1 above)
Comment (rarkins) [reply lib/config/definitions.ts]: I usually don’t mind if people opt into a feature they don’t understand and it doesn’t work as expected. But it depends on the failure case. I think if the feature is opt in and the risk is documented then I’m pretty unlikely to get complaints, maybe just a question or two.

File: lib/workers/pr/index.js (hunk 3 above)
Comment (rarkins) [reply lib/workers/pr/index.js]: But you’re not checking for config.allowMerge..... here. Or is it because it’s set at initRepo time and doesn’t need to be checked here also?

File: lib/config/definitions.ts (hunk 1 above)
Comment (fgreinacher) [reply lib/config/definitions.ts]: Alright, I’ll change it so that it can also be configured from within the repo!

File: lib/workers/pr/index.js (hunk 3 above)
Comment (fgreinacher) [reply lib/workers/pr/index.js]: Yes, it’s currently handled within the platform because it saw it more as a policy that only the platform should care about. But now that’s it’ll become a regular config option I can also add this check here and remove it from the platform.

File: lib/config/definitions.ts (hunk 2 above)
Comment (JamieMagee) [reply lib/config/definitions.ts]: The equivalent feature in Azure DevOps is called `azureAutoComplete`, so maybe `gitLabAutomerge`?

Comment (fgreinacher) [reply lib/config/definitions.ts]: In GitLab its actually called `merge when pipeline succeeds`, but that’s a bit lengthy, so I’ll go for `gitLabAutomerge` 😀

File: docs/usage/self-hosted-configuration.md [hunk 4]
```diff
@@ -7,6 +7,10 @@ description: Self-Hosted Configuration usable in renovate.json or package.json
 
 The below configuration options are applicable only if you are running your own instance ("bot") of Renovate.
 
+## allowMergeWhenPipelineSucceeds
+
+When this option is enabled it is possible that MRs with failing pipelines are getting merged. This is caused by a race condition in GitLab's Merge Request API - [read the issue](https://gitlab.com/gitlab-org/gitlab/issues/26293) for details.
```
Comment (JamieMagee) [inline docs/usage/self-hosted-configuration.md]: Can you also add some text about what it is supposed to do? I agree that the warning is necessary, but a sentence or two about the successful behaviour might be useful as well.

Comment (fgreinacher) [reply docs/usage/self-hosted-configuration.md]: The happy case is explained at https://github.com/renovatebot/renovate/pull/4558/files#diff-65aa9acc5e40f0995fc4d4ab67a2eeafR2004. The text here is just appended (please correct me when I got this wrong @rarkins).

File: lib/config/definitions.ts (hunk 1 above)
Comment (fgreinacher) [reply lib/config/definitions.ts]: Changed with ea783946d0d76f953c6fe3b4c3f21026d2318010

File: lib/config/definitions.ts (hunk 2 above)
Comment (fgreinacher) [reply lib/config/definitions.ts]: Done with ea783946d0d76f953c6fe3b4c3f21026d2318010

File: lib/workers/pr/index.js (hunk 3 above)
Comment (fgreinacher) [reply lib/workers/pr/index.js]: Done with ea783946d0d76f953c6fe3b4c3f21026d2318010

**Comment by fgreinacher** [comment]
OK, I think I addressed all comments in 7f2701cb6ca46f4ab6537a4112b93ca7cc54cc5e (please squash when merging)

File: lib/platform/gitlab/index.ts [hunk 5]
```diff
@@ -34,8 +39,8 @@ export async function initPlatform({
   endpoint,
   token,
 }: {
-  endpoint: string;
   token: string;
+  endpoint?: string;
```
Comment (rarkins) [inline lib/platform/gitlab/index.ts]: Why this change?

Comment (fgreinacher) [reply lib/platform/gitlab/index.ts]: Oops, unintentional 🧐 Reverted it.
